src/main/java/edu/umich/its/cpm/MigrationTaskService.java

```diff
@@ -1121,5 +1125,318 @@ private String getExistingBoxFolderIdFromBoxException(BoxAPIException e,
+	
+				downloadStatus = getMailArchiveZipContent(env, site_id, downloadStatus,
+						sessionId, httpContext, out);
+				
+				out.flush();
+				out.close();
+				log.info("Finished mail archive zip download for site " + site_id);
+				
+
+			} catch (RestClientException e) {
+				String errorMessage = Utils.STATUS_FAILURE + " Cannot find site by siteId: " + site_id
+						+ " " + e.getMessage();
+				Response.status(Response.Status.NOT_FOUND).entity(errorMessage)
+						.type(MediaType.TEXT_PLAIN).build();
+				log.error(errorMessage);
+				downloadStatus.put("status", errorMessage);
+			}
+			catch (IOException e) {
+				String errorMessage = Utils.STATUS_FAILURE + " Problem getting mail archive zip file for "
+						+ site_id + " " + e.getMessage();
+				Response.status(Response.Status.INTERNAL_SERVER_ERROR)
+						.entity(errorMessage).type(MediaType.TEXT_PLAIN)
+						.build();
+				log.error(errorMessage);
+				downloadStatus.put("status", errorMessage);
+
+			} catch (Exception e) {
+				String errorMessage = Utils.STATUS_FAILURE + " Migration status for " + site_id + " "
+						+ e.getClass().getName();
+				Response.status(Response.Status.INTERNAL_SERVER_ERROR)
+						.entity(errorMessage).type(MediaType.TEXT_PLAIN)
+						.build();
+				log.error("downloadMailArchiveZipFile ", e);
+				downloadStatus.put("status", errorMessage);
+			}
+		} else {
+			String userError = "Cannot become user " + userId;
+			log.error(userError);
+			downloadStatus.put("status", userError);
+		}
+
+		// update the status and end_time of migration record
+		setMigrationEndTimeAndStatus(migrationId, repository, downloadStatus);
+
+		return;
+
+	}
+
+	/**
+	 * get MailArchive message content into ZipOutputStream
+	 * @param env
+	 * @param site_id
+	 * @param downloadStatus
+	 * @param sessionId
+	 * @param httpContext
+	 * @param out
+	 * @return
+	 * @throws IOException
+	 */
+	private JSONObject getMailArchiveZipContent(Environment env, String site_id,
+			JSONObject downloadStatus, String sessionId,
+			HttpContext httpContext, ZipOutputStream out) throws IOException {
+
+		JSONArray messagesStatus = new JSONArray();
+		
+		// get all mail channels inside the site
+		RestTemplate restTemplate = new RestTemplate();
+		String requestUrl = env.getProperty(Utils.ENV_PROPERTY_CTOOLS_SERVER_URL)
+				+ "direct/mailarchive/siteChannels/" + site_id + ".json?_sessionId="
+				+ sessionId;
+		JSONObject channelsJSON = null;
+		channelsJSON = new JSONObject(restTemplate.getForObject(requestUrl,
+				String.class));
+		
+		if (!channelsJSON.has(Utils.JSON_ATTR_MAILARCHIVE_COLLECTION))
+		{
+			return downloadStatus;
+		}
+		
+		JSONArray channels = channelsJSON.getJSONArray(Utils.JSON_ATTR_MAILARCHIVE_COLLECTION);
+		
+		boolean folderForChannels = false;
+		if (channels.length() > 1)
+		{
+			// if the site have more than one MailArchive channel
+			// create zip folder for each channel
```

Comment:
This is a tradition data structure used by legacy Sakai code (chat, discussion, email archive, etc). The idea is that there could be different communication/email channels for one single site. But I just checked the CTools UI, there is no option for user to create different email channel in UI. So I guess all email messages in site will fall into the default channel - namely "main" channel.